numba_cuda/numba/cuda/dispatcher.py

```diff
@@ -262,22 +264,53 @@ def link_to_library_functions(
         self.reload_init = []
 
     def maybe_link_nrt(self, link, tgt_ctx, asm):
+        """
+        Add the NRT source code to the link if the neccesary conditions are met:
+        If NRT is enabled for the CUDATargetContext, return True if we either
+        detect .extern .func decls of NRT functions in any file in the link or
+        any of the passed LinkableCode objects have NRT enabled. In the special
+        case of PTXSource or CUSource objects, the source code will be inspected
+        even if the NRT flag is not set.
```

Comment:
Why are we inspecting the source of PTX and CUDA C/C++ source files? Shouldn't the user be expected to explicitly state that they require NRT by setting the `LinkableCode` `nrt` property? There's a couple of reasons I think it would be better to require this:

* I think a while back we concluded it would be hard to perfectly determine whether arbitrary linkable code used the NRT (potential false positives and false negatives), so we shouldn't try.
* A user might pass in an object as PTX or CUDA source, and rely on NRT getting linked in automatically. If they then decide to switch to LTO-IR for better performance, the automatic linking will no longer work, and their kernels will appear to suddenly have unresolved NRT symbols.

Reply:
I am quite +1 on requiring the user the be explicit about when they need NRT and less autodetection. Does the following seem right then? We'd link NRT if:

- it is globally enabled

**And** either:
- NRT is autodetected in the main asm (this is the only thing we try and autodetect NRT symbols in)
- a LinkableCode object with NRT specifically enabled is passed

If a user passes a `.cu` or `.ptx` file (in any form, either as a path or as a `LinkableCode` instance) _without_ NRT explicitly set, it may fail to link if the code does contain NRT symbols.

Reply:
Should be addressed in https://github.com/NVIDIA/numba-cuda/pull/167/commits/ff12e16cc0ede0d35e5f71b3afbcd2fc3db52ee4

Reply:
I think this is addressed from a technical standpoint (modulo the minor comment below) but the docstring here is now out of date, suggesting that PTX and CU sources will be inspected.